autox/informer.go

```diff
@@ -3,154 +3,420 @@ package autox
+			return
+		}
 
-	// check if the app name matches the name in the autox label
-	if autoxLabelName, ok := labels[autoxLabel]; ok && resourceName == autoxLabelName {
-		// Delete Helm charts
-		deleteHelmReleases(resourceName)
+		prunedLabels := pruneLabels(clientU.GetLabels())
+
+		// check if labels have changed
+		if reflect.DeepEqual(oldPrunedLabels, prunedLabels) {
+			log.Logger.Debug("updated object has no labels changes")
+			return
+		}
+
+		// if labels have changed, then delete the Helm releases
+		_ = deleteObjectHelper(clientU, prunedLabels, autoXConfig)
+
+		// check if autoX label exists
+		labels := clientU.GetLabels()
+		if !hasAutoXLabel(labels) {
+			log.Logger.Warnf("expected object \"%v\" to have label \"%s\" but none were found", clientU, autoXLabel)
+			return
+		}
+
+		// only install Helm releases if autoX label exists
+		_ = addObjectHelper(clientU, prunedLabels, autoXConfig)
 	}
 }
 
+func deleteObject(ns string, gvr schema.GroupVersionResource, m *sync.Mutex, autoXConfig config) func(obj interface{}) {
+	return func(obj interface{}) {
+		m.Lock()
+		defer m.Unlock()
+
+		log.Logger.Debug("Delete:", obj)
+
+		u := obj.(*unstructured.Unstructured)
+		clientU, _ := k8sClient.dynamicClient.Resource(gvr).Namespace(ns).Get(context.TODO(), u.GetName(), metav1.GetOptions{})
+		if clientU != nil {
+			log.Logger.Warnf("object \"%v\" should have been deleted but was not", clientU)
+			return
+		}
+
+		// delete Helm releases if autoX label exists
+		labels := u.GetLabels()
+		prunedLabels := pruneLabels(labels)
+		_ = deleteObjectHelper(u, prunedLabels, autoXConfig)
+	}
+}
+
+func deleteObjectHelper(uObj *unstructured.Unstructured, prunedLabels map[string]string, autoXConfig config) error {
+	return deleteHelmReleases(prunedLabels, uObj.GetNamespace(), autoXConfig)
+}
+
 type iter8Watcher struct {
 	factories map[string]dynamicinformer.DynamicSharedInformerFactory
 }
 
-func newIter8Watcher() *iter8Watcher {
+func newIter8Watcher(autoXConfig config) *iter8Watcher {
 	w := &iter8Watcher{
 		factories: map[string]dynamicinformer.DynamicSharedInformerFactory{},
 	}
-	// for each namespace, resource type configure Informer
-	for _, ns := range iter8ResourceConfig.Namespaces {
+
+	// aggregate all triggers (namespaces and GVR) from the releaseGroupConfig
+	// triggers map has namespace as its key and the object GVRs within the namespace that it is watching as its value
+	triggers := map[string][]schema.GroupVersionResource{}
+	for _, releaseGroupSpec := range autoXConfig.Specs {
+
+		namespace := releaseGroupSpec.Trigger.Namespace
+		gvr := schema.GroupVersionResource{
+			Group:    releaseGroupSpec.Trigger.Group,
+			Version:  releaseGroupSpec.Trigger.Version,
+			Resource: releaseGroupSpec.Trigger.Resource,
+		}
+
+		// add namespace and GVR to triggers
+		triggers[namespace] = append(triggers[namespace], gvr)
+	}
+
+	// mutex is passed to event handler so changes are not concurrent
+	var m sync.Mutex
```

Comment:
Make mutex package local variable (so it is accessible anywhere inside the package).